File: Sprint5Exercises/JsExercises/BiscuitMaker.js
class Recipe {
    constructor(name, flour, butter, soda, sugar) {
        if (flour < 0 || butter < 0 || soda < 0 || sugar < 0) {
            throw new Error("Ingredient amounts cannot be negative.");
        }
        this.name = name;
        this.flour = flour;
        this.butter = butter;
        this.soda = soda;
        this.sugar = sugar;
    }
}

class Solution {
    constructor(flour, butter, soda, sugar) {
        this.biscuits = [];
        this.remainingFlour = flour;
        this.remainingButter = butter;
        this.remainingSoda = soda;
        this.remainingSugar = sugar;
    }

    addBiscuit(biscuitName) {
        this.biscuits.push(biscuitName);
    }

    toString() {
        const temp = JSON.stringify(this.getUniqueCounts(this.biscuits));
        return temp + ` Remaining Ingredients [Flour: ${this.remainingFlour}, Butter: ${this.remainingButter}, Soda: ${this.remainingSoda}, Sugar: ${this.remainingSugar}]` + "\n";
    }

    getUniqueCounts(biscuits){
        const uniqueElements = {};
        biscuits.forEach((element) => {
        if (!uniqueElements.hasOwnProperty(element)) {
            uniqueElements[element] = 0;
        }
        uniqueElements[element]++;
        });
        return uniqueElements;
    }
}

const biscuitRecipes = [
    new Recipe("Circular", 1, 2, 0, 3),
    new Recipe("Triangular", 2, 1, 1, 2),
    new Recipe("Rectangular", 3, 1, 2, 1)
];

const findCombinations = (flour, butter, soda, sugar, current, solutions) => {
    if (flour < 0 || butter < 0 || soda < 0 || sugar < 0) {
        throw new Error("Ingredient amounts cannot be negative.");
    }
    let added = false;
    for (let recipe of biscuitRecipes) {
        if (flour >= recipe.flour && butter >= recipe.butter &&
            soda >= recipe.soda && sugar >= recipe.sugar) {
            let newSolution = new Solution(flour - recipe.flour,
                                           butter - recipe.butter,
                                           soda - recipe.soda,
                                           sugar - recipe.sugar);
            newSolution.biscuits = [...current.biscuits];
            newSolution.addBiscuit(recipe.name);

            findCombinations(newSolution.remainingFlour, newSolution.remainingButter,
                             newSolution.remainingSoda, newSolution.remainingSugar,
                             newSolution, solutions);
            added = true;
        }
    }
    if (!added && current.biscuits.length > 0) {
        solutions.push(current);
    }
}

const main = () => {
    try {
        let solutions = [];
        findCombinations(10, 16, 17, 9, new Solution(10, 16, 17, 9), solutions);
        for (let sol of solutions) {
            console.log(sol.toString());
        }
        console.log(solutions.length);
    } catch (e) {
        console.error("Error: " + e.message);
    }
}

main();

File: Sprint5Exercises/JsExercises/BiscuitMaker.java
import java.util.ArrayList;
import java.util.List;

public class BiscuitMaker {
    static class Recipe {
        int flour, butter, soda, sugar;

        public Recipe(int flour, int butter, int soda, int sugar) {
            if (flour < 0 || butter < 0 || soda < 0 || sugar < 0) {
                throw new IllegalArgumentException("Ingredient amounts cannot be negative.");
            }
            this.flour = flour;
            this.butter = butter;
            this.soda = soda;
            this.sugar = sugar;
        }
    }

    static class Biscuit {
        String name;
        Recipe recipe;

        public Biscuit(String name, Recipe recipe) {
            this.name = name;
            this.recipe = recipe;
        }
    }

    static class Solution {
        List<String> biscuits = new ArrayList<>();
        int remainingFlour, remainingButter, remainingSoda, remainingSugar;

        public Solution(int flour, int butter, int soda, int sugar) {
            this.remainingFlour = flour;
            this.remainingButter = butter;
            this.remainingSoda = soda;
            this.remainingSugar = sugar;
        }

        public void addBiscuit(String biscuit) {
            biscuits.add(biscuit);
        }

        @Override
        public String toString() {
            return biscuits.toString() + " Remaining Ingredients [Flour: " + remainingFlour + ", Butter: " + remainingButter + ", Soda: " + remainingSoda + ", Sugar: " + remainingSugar + "]";
        }
    }

    static Recipe[] recipes = {
            new Recipe(1, 2, 0, 3), // Circular
            new Recipe(2, 1, 1, 2), // Triangular
            new Recipe(3, 1, 2, 1)  // Rectangular
    };
    static String[] biscuitNames = {"Circular", "Triangular", "Rectangular"};

    public static void findCombinations(int flour, int butter, int soda, int sugar, Solution current, List<Solution> solutions) {
        if (flour < 0 || butter < 0 || soda < 0 || sugar < 0) {
            throw new IllegalArgumentException("Ingredient amounts cannot be negative.");
        }
        boolean added = false;
        for (int i = 0; i < recipes.length; i++) {
            if (flour >= recipes[i].flour && butter >= recipes[i].butter &&
                soda >= recipes[i].soda && sugar >= recipes[i].sugar) {
                Solution newSolution = new Solution(flour - recipes[i].flour,
                                                    butter - recipes[i].butter,
                                                    soda - recipes[i].soda,
                                                    sugar - recipes[i].sugar);
                newSolution.biscuits.addAll(current.biscuits);
                newSolution.addBiscuit(biscuitNames[i]);

                findCombinations(newSolution.remainingFlour, newSolution.remainingButter,
                                 newSolution.remainingSoda, newSolution.remainingSugar,
                                 newSolution, solutions);
                added = true;
            }
        }
        if (!added && !current.biscuits.isEmpty()) { // Ensure we only add solutions that actually produced biscuits
            solutions.add(current);
        }
    }

    public static void main(String[] args) {
        try {
            List<Solution> solutions = new ArrayList<>();
            findCombinations(10, 16, 17, 9, new Solution(10, 16, 17, 9), solutions); // Example ingredients
            for (Solution sol : solutions) {
                System.out.println(sol);
            }
        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("An unexpected error occurred: " + e.getMessage());
        }
    }
}
